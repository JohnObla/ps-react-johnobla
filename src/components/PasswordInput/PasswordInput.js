import React from 'react';
import PropTypes from 'prop-types';
import useToggleBool from '../../hooks/useToggleBool';
import ProgressBar from '../ProgressBar';
import EyeIcon from '../EyeIcon';
import TextInput from '../TextInput';
import './PasswordInput.scss';

/** Password inpuit with integratedlabel, quality tips, and show password toggle. */
const PasswordInput = ({
  htmlId,
  value,
  label = 'Password',
  error,
  onChange,
  placeholder,
  maxLength = 50,
  showPasswordInitially = false,
  quality,
  ...props
}) => {
  const [showPassword, toggleShowPassword] = useToggleBool(
    showPasswordInitially
  );

  const handleToggle = event => {
    toggleShowPassword();
    if (event) event.preventDefault();
  };

  return (
    <TextInput
      htmlId={htmlId}
      label={label}
      placeholder={placeholder}
      type={showPassword ? 'text' : 'password'}
      onChange={onChange}
      value={value}
      maxLength={maxLength}
      error={error}
      required
      {...props}
    >
      {showPassword && (
        <a className="PasswordInput__toggle" href="#" onClick={handleToggle}>
          <EyeIcon />
        </a>
      )}
      {value.length > 0 && quality && (
        <ProgressBar percent={quality} width={130} />
      )}
    </TextInput>
  );
};

PasswordInput.propTypes = {
  /** Unique HTML ID. Used for tying label to HTML input. Handy hook for automated testing. */
  htmlId: PropTypes.string.isRequired,

  /** Input name. Recommend setting this to match object's property so a single change handler can be used*/
  name: PropTypes.string.isRequired,

  /** Password Value */
  value: PropTypes.any,

  /** Input label */
  label: PropTypes.string.isRequired,

  /** Function called when password input value changes */
  onChange: PropTypes.func.isRequired,

  /** Max password length accepted */
  maxLength: PropTypes.number,

  /** Placeholder to display when no password is entered */
  placeholder: PropTypes.string,

  /** When set to true, shows the toggle for displaying the currently entered password */
  showPasswordInitially: PropTypes.bool,

  /** Displaypassword quality visually via ProgressBar, accepts a number between 0 and 100 */
  quality: PropTypes.number,

  /** Validation error to display */
  error: PropTypes.string,
};

export default PasswordInput;
