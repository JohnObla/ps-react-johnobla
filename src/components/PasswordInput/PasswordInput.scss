.PasswordInput__toggle {
  margin-left: 5px;
}
