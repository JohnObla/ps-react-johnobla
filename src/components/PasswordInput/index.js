import './PasswordInput.scss';
export { default } from './PasswordInput';
