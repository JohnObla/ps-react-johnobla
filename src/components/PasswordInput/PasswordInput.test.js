import React from 'react';
import renderer from 'react-test-renderer';
import PasswordInput from './PasswordInput';

describe('PasswordInput', () => {
  it('should hide password quality by default', () => {
    const tree = renderer
      .create(
        <PasswordInput
          htmlId="test"
          name="test"
          onChange={() => {}}
          value="This Is a Dummy Password"
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
