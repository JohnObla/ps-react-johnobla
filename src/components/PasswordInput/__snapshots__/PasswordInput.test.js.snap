// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput should hide password quality by default 1`] = `
<div
  className="TextInput"
>
  <label
    className="Label"
    htmlFor="test"
  >
    Password
     
    <span
      className="Label__required"
    >
       *
    </span>
  </label>
  <input
    className="TextInput__inputundefined"
    id="test"
    maxLength={50}
    name="test"
    onChange={[Function]}
    type="password"
    value="This Is a Dummy Password"
  />
  <button
    className="PasswordInput__toggle"
    onClick={[Function]}
  >
    <svg
      height="16"
      viewBox="0 0 22 22"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="#4d4d4d"
        transform="matrix(.02146 0 0 .02146 1 1)"
      >
        <path
          d="m466.07 161.53c-205.6 0-382.8 121.2-464.2 296.1-2.5 5.3-2.5 11.5 0 16.9 81.4 174.9 258.6 296.1 464.2 296.1 205.6 0 382.8-121.2 464.2-296.1 2.5-5.3 2.5-11.5 0-16.9-81.4-174.9-258.6-296.1-464.2-296.1m0 514.7c-116.1 0-210.1-94.1-210.1-210.1 0-116.1 94.1-210.1 210.1-210.1 116.1 0 210.1 94.1 210.1 210.1 0 116-94.1 210.1-210.1 210.1"
        />
        <circle
          cx="466.08"
          cy="466.02"
          r="134.5"
        />
      </g>
    </svg>
  </button>
</div>
`;
